x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -130,6 +136,44 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeMap(metadata);
     }
 
+    /**
+     * Checks whether the user or API key of the passed in authentication can access the resources owned by the user
+     * or API key of this authentication. The rules are as follows:
+     *   * True if the authentications are for the same API key (same API key ID)
+     *   * True if they are the same username from the same realm
+     *      - For file/native/kerberos realm, same realm means the same realm type
+     *      - For all other realms, same realm means same realm type plus same realm name
+     *   * An user and its API key cannot access each other's resources
+     *   * An user and its token can access each other's resources
+     *   * Two API keys are never able to access each other's resources regardless of their ownership.
+     *
+     *  This check is a best effort and it does not account for certain static and external changes.
+     *  See also <a href="https://www.elastic.co/guide/en/elasticsearch/reference/master/security-limitations.html">
+     *      security limitations</a>
+     * @param other
+     * @return
+     */
+    public boolean canAccessResourcesOf(Authentication other) {
+        if (AuthenticationType.API_KEY == getAuthenticationType() && AuthenticationType.API_KEY == other.getAuthenticationType()) {
+            final boolean sameKeyId = getMetadata().get(API_KEY_ID_KEY).equals(other.getMetadata().get(API_KEY_ID_KEY));
+            if (sameKeyId && Assertions.ENABLED) {
+                assert getUser().principal().equals(other.getUser().principal()) :
+                    "The same API key ID cannot be attributed to two different usernames";
+            }
+            return sameKeyId;
+        }
+        if (false == getUser().principal().equals(other.getUser().principal())) {
+            return false;
+        }
```

Comment:
I think we need to be defensive about `AuthenticationType` here.

The javadocs state that we want to handle `REALM` and `TOKEN` as being a match, but we should probably reject anything else, or at least reject anything else that aren't identical on both authentications.

No matter what we do, I think we need an assert that will fail if we get an unexpected (newly created) authentication type. This code doesn't know how to handle `BIOMETRICS`, and there's no reason to believe that the current implementation would be correct. I think we want to fail fast if we end up here with an auth type we don't support.

Reply:
Nice catch. The logic is indeed implicit here and is not consistent with what the javadoc describes. I agree that `REALM` and `TOKEN` are the real use case where two different authentication types should be considered equal. `ANONYMOUS` is of its own type and any `INTERNAL` authenticated users should not ever be authenticated by other authentication types (it would be a bug if they do). I ended-up with following rules:
* Two identical authentication types always pass the "can-share-resources" check including any possible future new authentication types (the authentication objects do still subject to other checks e.g. username, realm).
* Two different authentication types can only pass "can-share-resources" check if one of them is `REALM` and the other is `TOKEN`. Otherwise, it will fail the check.
* It's an assertion error if there is any unknown (new) AuthenticationType in the two different authentication types because whether two authentication types should be considered equal should be an explicit manual decision process.